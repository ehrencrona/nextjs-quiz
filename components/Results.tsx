import Question from "../model/Question";
import theme from "../styles/theme";

const Results = ({
  questions,
  answers,
  onRestart,
}: {
  questions: Question[];
  answers: Record<number, boolean>;
  onRestart: () => void;
}) => {
  const isCorrect = (questionNumber: number) =>
    questions[questionNumber].correct_answer === answers[questionNumber];

  const correctCount = questions.filter((question, questionNumber) =>
    isCorrect(questionNumber)
  ).length;

  return (
    <div>
      <h2>
        You scored {correctCount} out of {questions.length}
      </h2>

      {questions.map((question, questionNumber) => (
        <div
          key={questionNumber}
          className={
            "answer " + (isCorrect(questionNumber) ? "correct" : "incorrect")
          }
          dangerouslySetInnerHTML={({__html:question.question})}>
        </div>
      ))}

      <button onClick={onRestart}>Play again</button>

      <style jsx>{`
        .question {
          width: ${theme.pageWidth};
          max-width: 100%;
        }

        h2 {
          font-size: 1.6rem;
          letter-spacing: 0.01em;
          margin-bottom: 0.6em;
        }

        .answer {
          margin-bottom: 0.6em;
        }

        .answer::before {
          position: relative;
          display: inline-block;
          width: 0;
          left: -0.6em;
          font-weight: 600;
        }

        .correct {
          color: #2ca007;
        }

        .correct::before {
          content: "+";
        }

        .incorrect {
          color: #e24406;
        }

        .incorrect::before {
          content: "-";
        }

        .number {
          font-size: 0.8rem;
        }

        .buttons button {
          margin-right: 1em;
        }
      `}</style>
    </div>
  );
};

export default Results;
