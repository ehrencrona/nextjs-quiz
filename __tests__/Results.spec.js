import { render } from "@testing-library/react";
import React from "react";
import Results from "../components/Results";

describe("result page", () => {
  test("shows correct answers", async () => {
    const question = {
      question: "We are living in a simulation.",
      category: "Metaphysics",
      correct_answer: true
    }

    const { getByText, container } = render(
      <Results
        questions={[question, question, question]}
        answers={{0: true, 1: false, 2:true}}
      />
    );

    getByText("You scored 2 out of 3");

    expect(container).toMatchSnapshot()
  });
});
