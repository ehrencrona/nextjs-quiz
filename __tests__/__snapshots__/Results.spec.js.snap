// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`result page shows correct answers 1`] = `
<div>
  <div
    class="jsx-1356833245 page"
  >
    <div
      class="jsx-1356833245 content"
    >
      <h2
        class="jsx-1915113496"
      >
        You scored 
        2
         out of 
        3
      </h2>
      <div
        class="jsx-1915113496 answer correct"
      >
        We are living in a simulation.
      </div>
      <div
        class="jsx-1915113496 answer incorrect"
      >
        We are living in a simulation.
      </div>
      <div
        class="jsx-1915113496 answer correct"
      >
        We are living in a simulation.
      </div>
    </div>
    <div
      class="jsx-1356833245 buttons"
    >
      <div
        class="jsx-1356833245 fade"
      />
      <button
        class="jsx-1356833245"
      >
        Play again
      </button>
    </div>
  </div>
</div>
`;
